torch/fx/experimental/migrate_gradual_types/constraint_transformation.py

```diff
@@ -0,0 +1,724 @@
+            return Disj(disj), counter
+        else:
+            return constraint, counter
+
+    # dimension constraints
+    elif isinstance(constraint, BinConstraintD):
+
+        if constraint.op == op_precision:
+            if isinstance(constraint.lhs, int):
+                return BinConstraintD(constraint.lhs, constraint.rhs, op_eq), counter
+            elif constraint.lhs == Dyn:
+                return T(), counter
+
+        elif constraint.op == op_consistency:
+            return Disj([BinConstraintD(constraint.lhs, constraint.rhs, op_eq),
+                         BinConstraintD(constraint.rhs, Dyn, op_eq), BinConstraintD(constraint.lhs, Dyn, op_eq)]), counter
+
+        else:
+            return constraint, counter
+
+    elif isinstance(constraint, Conj):
+        new = []
+        for c in constraint.conjucts:
+            new_c, counter = transform_constraint(c, counter)
+            new.append(new_c)
+        return Conj(new), counter
+
+    elif isinstance(constraint, Disj):
+        new = []
+        for c in constraint.disjuncts:
+            new_c, counter = transform_constraint(c, counter)
+            new.append(new_c)
+        return Disj(new), counter
+
+    # this is up to size 1 and 2 only for now
+    elif isinstance(constraint, TGreatestUpperBound):
+        c1 = Conj([Disj([BinConstraintT(constraint.rhs1, Dyn, op_eq),
+                         BinConstraintT(constraint.rhs2, Dyn, op_eq)]), BinConstraintT(constraint.res, Dyn, op_eq)])
+
+        [c2, c3, c4, c5], counter = gen_greatest_upper_bound(constraint, counter)
+
+        return Disj([c1, c2, c3, c4, c5]), counter
+
+    elif isinstance(constraint, DGreatestUpperBound):
+        c1 = Conj([BinConstraintD(constraint.rhs1, Dyn, op_eq), BinConstraintD(constraint.res, constraint.rhs2, op_eq)])
+        c2 = Conj([BinConstraintD(constraint.rhs2, Dyn, op_eq), BinConstraintD(constraint.res, constraint.rhs1, op_eq)])
+        c3 = Conj([BinConstraintD(constraint.rhs2, constraint.rhs1, op_eq), BinConstraintD(constraint.res, constraint.rhs1, op_eq)])
+        return Disj([c1, c2, c3]), counter
+
+    elif isinstance(constraint, CalcConv):
+        transformed, counter = transform_conv(constraint, counter)
+        return transformed, counter
+
+    elif isinstance(constraint, CalcMaxPool):
+        transformed, counter = transform_maxpool(constraint, counter)
+        return transformed, counter
+
+    elif isinstance(constraint, CalcProduct):
+        transformed, counter = transform_flatten(constraint, counter)
+        return transformed, counter
+
+    elif isinstance(constraint, CanReshape):
+        transformed, counter = transform_can_reshape(constraint, counter)
+        return transformed, counter
+
+    elif isinstance(constraint, ApplyBroadcasting):
+        transformed, counter = transform_apply_broadcasting(constraint, counter)
+        return transformed, counter
+    else:
+        return constraint, counter
```

Comment:
Can we turn this if/else tree into a registration pattern like in `constraint_generator.py`? It's also kind-of strange to be directly dispatching to another function for half of the constraint types (`CalcConv` etc) but then have logic embedded in the if/else for others